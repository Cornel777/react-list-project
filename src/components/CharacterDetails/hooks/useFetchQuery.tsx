import { useQuery, keepPreviousData } from "@tanstack/react-query";
import fetchSwapi from "../../../services/api/fetchSwapi";

export interface HookReturn<T> {
  isFetching: boolean;
  isPlaceholderData: boolean;
  isPending: boolean;
  isError: boolean;
  error: Error | null;
  data: T | undefined;
}


const useFetchQuery = <T, >(queryKey: string, path: string, id: string | undefined): HookReturn<T> => {

  const { isFetching, isPlaceholderData, isPending, isError, error, data } = useQuery<T>({
    queryKey: [queryKey, id],
    queryFn: () => fetchSwapi(path),
    staleTime: 5000,
    placeholderData: keepPreviousData,
  })

    return {
      isFetching,
      isPlaceholderData,
      isPending,
      isError,
      error,
      data,
    }
}

export default useFetchQuery;