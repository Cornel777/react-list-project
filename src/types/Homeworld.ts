import { URL } from "./Character";

export interface Homeworld {
  name: string;
  rotation_period: string;
  orbital_period: string;
  diameter: string;
  climate: string;
  gravity: string;
  terrain: string;
  surface_water: string;
  population: string;
  residents: string[];
  films: URL[];
  created: string;
  edited: string;
  url: URL;
}

