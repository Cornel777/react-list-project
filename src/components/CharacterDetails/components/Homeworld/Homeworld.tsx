import { FunctionComponent } from "react"
import useFetchQuery from "../../hooks/useFetchQuery"
import { type Homeworld } from "../../../../types/Homeworld"

interface HomeworldProps {
  homeworldUrl: string
}

const Homeworld: FunctionComponent<HomeworldProps> = ({homeworldUrl}) => {

  const homeworldUrlList = homeworldUrl.split('/').filter((element: string) => element.length > 0)
  const homeworldId = homeworldUrlList.pop()

  const {isPending, isError, error, data} = useFetchQuery<Homeworld>('homeworld', `planets/${homeworldId}/`, homeworldId)

  if (isPending) {
    return <div>Loading...</div>;
  }
  if (isError) {
    return <p>{`Failed to load homeworld ${error && error.message}`}</p>;
  }
  if (!data) {
    return null;
  }

  return data.name
}

export default Homeworld;