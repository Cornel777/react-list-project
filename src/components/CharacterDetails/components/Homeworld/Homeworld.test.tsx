import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import Homeworld from './Homeworld';
import useFetchQuery from '../../hooks/useFetchQuery';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

vi.mock('../../hooks/useFetchQuery');

const mockHomeworldData = {
  name: 'Tatooine',
  rotation_period: '23',
  orbital_period: '304',
  diameter: '10465',
  climate: 'arid',
  gravity: '1 standard',
  terrain: 'desert',
  surface_water: '1',
  population: '200000',
  url: 'https://swapi.dev/api/planets/1/'
};

describe('Homeworld', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false
      }
    }
  });

  beforeEach(() => {
    vi.clearAllMocks();
    vi.mocked(useFetchQuery).mockReturnValue({
      isPending: false,
      isError: false,
      isFetching: false,
      isPlaceholderData: false,
      error: null,
      data: mockHomeworldData
    });
  });

  function renderWithClient(ui: React.ReactElement) {
    const testQueryClient = queryClient;
    const { rerender, ...result } = render(
        <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>
    )
    return {
        ...result,
        rerender: (rerenderUi: React.ReactElement) =>
            rerender(
                <QueryClientProvider client={testQueryClient}>{rerenderUi}</QueryClientProvider>
            ),
    }
  }

  it('matches the snapshot', () => {
    const { asFragment } = renderWithClient(<Homeworld homeworldUrl="https://swapi.dev/api/planets/1/" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders homeworld details', () => {
    renderWithClient(<Homeworld homeworldUrl="https://swapi.dev/api/planets/1/" />);
    expect(screen.getByText('Tatooine')).toBeInTheDocument();
    expect(useFetchQuery).toHaveBeenCalledWith('homeworld', 'planets/1/', '1');
  });
});