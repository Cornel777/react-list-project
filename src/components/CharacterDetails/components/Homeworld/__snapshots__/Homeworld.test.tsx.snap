// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Homeworld > matches the snapshot 1`] = `
<DocumentFragment>
  Tatooine
</DocumentFragment>
`;
